### ISSUE TITLE
VMP: sending custom messages?

### ISSUE THREAD

**Issue by avonmoll** [bayespy/bayespy#75]
I have a graph that involves several Gaussian processes. One of my latent nodes, a_j, is separated from the next latent node by a deterministic node and a factor. I believe a_j requires a custom message. Is there a way to supply this message manually? Otherwise, what is the recommended way to update this particular node?
![gprn-factor-graph](https://cloud.githubusercontent.com/assets/4260850/14652990/d2dfe922-0644-11e6-8b6e-bdff1b285aaa.png)

**Comment by jluttine** [comment]
Thanks for asking, and thanks for the very clear and nice graph! :+1:

Do I understand correctly: `a_j` is gamma distributed and `K_{f_j}` is a covariance matrix and you'd like to scale the (inverse) covariance matrix by the gamma distributed scalar? Is it okay to compute the inverse covariance matrix and use that in the messages? I should implement a node which computes products of gamma and Wishart variables, that'd solve it. It is quite easy to implement, so I can do it soon.

Another issue is with the GP covariance matrices and hyperparameters. If you'd like to learn the hyperparameters of the covariance matrices, it's not yet supported, unfortunately. Also, currently you'd need to pass the inverse covariance matrix which might be something you don't want to do. I have plans to add GP support via GPy package in the near future, that would help a lot. But in that case, you might need to put `a_j` to the set of the hyperparameters and learn a ML point estimate for it. I'm not sure is it easily possible to add support for gamma distributed scalings with GPy. Might not be an issue, but I'm not yet sure.

For the other parts, I think the model is easy to implement in BayesPy.

**Comment by avonmoll** [comment]
Great to hear that specific support for GP's is already being looked at! I must admit, I'm still fairly new to probabilistic programming in general. Basically, all I am trying to do is implement (in an open source fashion) a technique called Gaussian Process Regression Networks ([paper](http://arxiv.org/pdf/1110.4411)). The authors used Infer.NET and they mentioned that they had to make modifications in order to implement the update for  `a_j`. I should also mention that they parameterized using variance instead of precision - thus they specify an INVERSE gamma prior for the variances. Doing so alleviates the need to invert the covariance matrices for the purposes of this graph. However, as I understand, there is currently no inverse gamma nodes in bayespy.

My thought for learning the hyperparameters of the covariance matrices was to treat them as constants, then update the latent nodes using VMP, then recomputing `\theta_f` and `\theta_W` using gradient descent to optimize the variational lower bound. Does this seem reasonable/possible?

Thanks so much for your response and consideration!

**Comment by jluttine** [comment]
Your approach definitely sounds reasonable. However, support for general maximum likelihood learning in BayesPy is still lacking. Yep, I know, shocking.. But I've been thinking how to implement it properly so that general non-conjugate black-box inference and type-II ML estimation all fit nicely into the same framework as they have similar requirements. And in your case, you want type-II ML estimation because the latent function values are strongly coupled with the covariance hyperparameters.

I'll take a look on this model with better time some day soon. I think the easiest/quickest way to get it implemented is by adding GPy support and then putting `a_j` as a hyperparameter which is point-estimated. But I need to think a bit. I'll get back to you.

**Comment by jluttine** [comment]
Just for your information: I'm working on the ML support and then the scaling. I hope I'll have time to finish it by the weekend. I think that should be sufficient to give you preliminary support for the model. GPy support is a bit bigger task, I'll do that some other time.

**Comment by jluttine** [comment]
I quickly implemented the necessary new features for your model: 1) scaling of Wishart node with Gamma node and 2) maximum likelihood estimation (needed for GP hyperparameters). I want to emphasize that MLE is very experimental and it is not efficiently implemented. For instance, the optimization algorithms aren't good. But as I'm planning to rewrite the main core of the whole system to use Theano/Tensorflow/Autograd, that rewrite will eventually fix that and lots of other issues in a much better way. That's why I didn't start to make any really elegant and comprehensive solutions because they'd be temporary anyway. For instance, MLE is supported for parameterized Wishart type variables only. It is easy to add support for any other nodes but it requires implementing some gradients so I didn't bother doing it now because they'd go to trash soon anyway.

That said, here's a simple GP regression example which uses both new features. Note that you need to use the develop branch. You need to implement the gradient passing functions yourself, because automatic/symbolic differentiation isn't used (yet). Probably this raises questions and I'm sure there are unclear parts. So please ask if you have further questions and interest in this.


```python
import bayespy as bp
from scipy.spatial import distance


# Some 3-dimensional inputs
N = 200
X = np.random.randn(N, 1)


def dist(X):
    return distance.squareform(distance.pdist(X))


# Squared distance matrix
D = dist(X) ** 2


def exponentiated_square(parameters):
    lengthscale = np.exp(parameters[0])
    return np.exp(-D/lengthscale) + 1e-6 * np.identity(N)


def d_parameters(d_exponentiated_square, parameters):
    """ Backward gradient of exponentiated square w.r.t. parameters """
    lengthscale = np.exp(parameters[0])
    K = np.exp(-D/lengthscale)
    return [
        np.sum(d_exponentiated_square * K * D / lengthscale),
    ]


def inverse(K):
    return np.linalg.inv(K)


def d_covariance(d_inverse, K):
    """ Backward gradient of inverse w.r.t. covariance """
    invK = np.linalg.inv(K)
    return -np.dot(invK, np.dot(d_inverse, invK))


parameters = bp.nodes.MaximumLikelihood(np.log([1]))
Covariance = bp.nodes.Function(
    exponentiated_square,
    (parameters, d_parameters)
)
Lambda = bp.nodes.Function(
    inverse,
    (Covariance, d_covariance)
)

magnitude = bp.nodes.Gamma(1e-3, 1e-3)

scaled_Lambda = bp.nodes.Wishart.scale(Lambda, magnitude, ndim=1)

# Latent Gaussian process
latent = bp.nodes.Gaussian(np.zeros(N), scaled_Lambda)

# Observation noise precision
tau = bp.nodes.Gamma(1e-3, 1e-3)

# Observations node
Y = bp.nodes.GaussianARD(latent, tau)

# Create data
K = 10**2 * exponentiated_square(np.log([0.3]))
data = bp.nodes.Gaussian(np.zeros(N), np.linalg.inv(K + 0.1**2*np.identity(N))).random()
Y.observe(data)

# Run inference:
Q = bp.inference.VB(Y, latent, magnitude, tau, parameters)
Q.optimize(parameters, magnitude, tau, collapsed=[latent], maxiter=200, verbose=False)

print("Learned GP lengthscale:", np.exp(parameters.get_moments()[0]))
print("Learned GP magnitude:", magnitude.get_moments()[0] ** (-0.5))
print("Learned noise std:", tau.get_moments()[0] ** (-0.5))
```

**Comment by avonmoll** [comment]
Thanks for all the work you've done on this! It appears that these new features fully address the needs of the original problem. I'd like to understand everything better before I attempt to tackle this and actually implement GPRNs. Here are a few questions I have at this point:

1. What does it mean for a variable to be collapsed (in the Q.optimize call)?
    1. I'm assuming that it just means that these nodes are functions of the variables being optimized and must be updated as a result
1. What is the relationship between lambda and scaled lambda (more specifically: why is it a Wishart .scale object)?
    1. I'm assuming this was part of the functionality that was added, but why was it necessary?
    1. What does the covariance matrix have to do with Wishart nodes?
    1. Does scaled_Lambda correspond to the uppermost `\times` block in my graph?
1. `d_parameters` computes the backward gradient of `exponentiated_square` w.r.t. parameters, so what is the meaning of the `d_exponentiated_square` parameter and where does it come from?
    1. If `exponentiated_square = exp(-D/lengthscale) + 1e-6*I`, then wouldn't the derivative w.r.t. `lengthscale` simply be <pre>d_parameters = exp(-D/lengthscale)*D/lengthscale<b><i>***2</i></b></pre>?

**Comment by jluttine** [comment]
1. In general, collapsing means that you integrate out some variables when optimizing some others. This is useful when there's a very strong coupling between the variables. In this case, this particular method was implemented: http://arxiv.org/abs/1206.5162. Otherwise learning is very slow because the latent function values determine the hyperparameters very strongly. From implementation point of view, this just means that whenever the value of the optimized variables are changed, the collapsed variables are updated by using standard updating. This adds extra computational cost so it's not always useful. But it corresponds to a collapsed inference described in the paper.

2. Scaled Lambda has been multiplied by magnitude variable so it is the product of Lambda and magnitude. It might be useful to understand the basics of the variational message passing (VMP) framework: Each node corresponds to a function or distribution. These are written in such a way that they are functions of sufficient statistics (let's call them moments now). So, it is important what kind of moments are expected from the parent nodes. For instance, Gaussian node expects "Gaussian" moments for the mean and "Wishart" moments for the precision. Thus, the covariance matrix needs to mapped to "Wishart" moments. This is basically done automatically and you don't need to worry about it. But now that you want to multiply Lambda with a scalar, one has to perform this operation in such a way that it is compatible with the Wishart moments required by the Gaussian node. So I wrote a scale function which takes Wishart moments and gamma moments as inputs and computes the product as Wishart moments. Wishart class is just a convenient placeholder for this function so that Wishart nodes can use it too. And yes, scaled_Lambda is basically the uppermost product block in your graph.

3. Consider a function `L(x(y(z)))`. The gradient is computed by using the chain rule `dL/dz = dL/dx * dx/dy * dy/dz`. The backward passing means that first `dL/dx` is computed and passed to the next gradient function which uses it to compute `dL/dy` which is again passed to the next function which then computes the final result `dL/dz`. This kind of backpropagation is useful when the differentiated function `L` is a scalar as is the case now (VB lower bound). So `d_exponentiated_square` is the gradient passed from the outer function, in this case it is what `d_covariance` outputs. I think I named those functions and variables a bit badly.. Anyway, deriving those gradient formulas with matrices can be a bit painful, sorry about that. Theano/Tensorflow/Autograd will solve that. They derive the gradients and do all the backpropagation stuff automatically. Until BayesPy uses those, you need to derive them yourself...

  i. Nope, the derivative isn't computed with respect to `lengthscale`, it is computed with respect to `parameters[0]` (which is the lengthscale in log-scale). So you need to use the derivative of `d lengthscale / d parameters[0]` as well and that is just `lengthscale` because `d exp(x) = exp(x)`. Thus, the squaring of the denominator cancels out and you just divide by `lengthscale` without the squaring. It's really easy to make mistakes with these. I'm not even 100% sure those gradients I quickly wrote are correct..

Did these answer your questions or is something still unclear?

If you try implementing GPRN with BayesPy, please report if you experience any issues. I'm almost sure that you will hit some bugs or missing features because this requires some experimental features.

And also, the optimization function is poor for this problem as I warned already. It was meant for the optimization problem described in the collapsed inference paper. It's probably really bad for general purpose optimization. But it should give you something to start with at least. It shouldn't be very difficult to write some wrapper functions and use SciPy's optimization tools.

**Comment by avonmoll** [comment]
That clears things up quite a bit - especially the response to question 3. I feel like I have a better sense of the inference process.

This is somewhat of a side project for me, so it will likely be slow goings, but I will report if I run into any more roadblocks or issues.

One more question, though. I am still somewhat puzzled by the `d_covariance` function. The comment states:
> Backward gradient of inverse w.r.t. covariance

So we have `inverse = K^-1`. However, in the return value, it appears to be the formula for the derivative of `inverse` w.r.t. a __scalar__ (see http://www.atmos.washington.edu/~dennis/MatrixCalculus.pdf, pg. 8,9). However, it seems like we're interested in the derivative of `inverse` w.r.t. the covariance (`K`), as in <br>
`d K^-1 / d K`, which has a different formula (according to http://www4.ncsu.edu/~pfackler/MatCalc.pdf): <br>
`-(K^-T * K^-1)` where `*` is the Kronecker product. However, this would mean the size of `d_inverse`'s return value would be `N^2 x N^2` which doesn't seem to make sense. I apologize if this is getting off-topic or beside the point.

**Comment by jluttine** [comment]
The comment in the code might be a bit confusing. Let's say you have a scalar output function `L` and you use it as `L(x(y(z)))`. `x`, `y` and `z` do not need to be scalars, and quite often they're not. Now you want to compute `dL / dz` and you apply the chain rule in backpropagation style as I described before. At the last step, you want to compute `dL / dz`. You are given `dL / dy` and `z`. Note that this is a gradient of a *scalar* function with respect to a non-scalar function. That's the point of the backpropagation: because the optimized function is a scalar function, you are passing just normal gradients, not matrix gradients with respect to matrix functions. So, we are not really interested in `d K^-1 / d K`, but `d L / dK = d L / d K^-1 * d K^-1 / d K`. And this chain rule can be computed efficiently by using the first matrix calculus link you provided without computing the individual terms like `d K^-1 / dK` explicitly. So it is important to remember that you are computing gradients of a *scalar* function and that's why the chain ruling becomes very efficient. If you are unfamiliar with derivatives in the matrix world and you'd like to understand them better, you might want to try to derive the gradient of, for instance, `L = x^T * K^-1 * x` w.r.t. `K` as an exercise. You'll notice that you don't need to compute `-(K^-T * K^-1)` Kronecker product explicitly.

About the code: so `d_covariance` isn't computing `d K^-1 / d K` but `d L / dK` given `d L / d K^-1` and `K`. `d_covariance` doesn't need to know what happens at the outer functions, it's just given `d L / d K^-1` and that tells everything it needs to know. Also, it doesn't compute `d K^-1 / d K` explicitly, it just applies the chain rules.

**Comment by avonmoll** [comment]
Ah that makes sense! I think what I was missing was the fact that the formula in those notes is given as the derivative of a matrix inverse w.r.t. a scalar, i.e.:

`dK^-1 / dL = -K^-1 dK/dL K^-1`

But this can simply be rearranged (by dividing each side by each of the derivative terms):

`dL / dK = -K^-1 dL / dK^-1 K^-1`

Which is __precisely__ what is written in the code. Until now, that small step escaped me and caused me to be confused. P.S. I wish you could write LaTeX in Github :smile:

**Comment by avonmoll** [comment]
I repeated the example 50 times (same input points, different observed data each run) and got the following results:
![image](https://cloud.githubusercontent.com/assets/4260850/18172849/306a4b5a-7035-11e6-8a93-107f9b19ee88.png)
These are pretty good! It seems like this feature will be suitable for use in implementing GPRN.

Now I have a question related to `MaximumLikelihood` nodes where you have plates. I checked the the code and it looks like `MaximumLikelihood` sets the plates shape to the match the shape of the constructor argument. However, in the original graph (above) what could I do about `\theta_f` whose shape is `(n,)` (where `n` is the number of parameters required to specify the kernel) and has `j` plates? The reason for `j` plates is because each latent `f` node has its own Gaussian Process representation.
